validator/src/main/java/com/amazon/aoc/validators/TraceValidator.java

```diff
@@ -19,97 +19,99 @@
-                      .collect(Collectors.toList()));
-
-          traceList.sort(Comparator.comparing(Trace::getId));
-
-          log.info("expectedTraceList: {}", expectedTraceList);
-          log.info("traceList got from backend: {}", traceList);
-          if (expectedTraceList.size() != traceList.size()) {
-            throw new BaseException(ExceptionCode.TRACE_LIST_NOT_MATCHED);
-          }
-
-          for (int i = 0; i != expectedTraceList.size(); ++i) {
-            Trace trace = traceList.get(i);
-            compareTwoTraces(expectedTraceList.get(i), trace);
-          }
-        });
-  }
-
-  private void compareTwoTraces(Trace trace1, Trace trace2) throws BaseException {
-    // check trace id
-    if (!trace1.getId().equals(trace2.getId())) {
-      throw new BaseException(ExceptionCode.TRACE_ID_NOT_MATCHED);
+    @Override
+    public void init(Context context, ICaller caller, FileConfig expectedTrace) throws Exception {
+        this.xrayService = new XRayService(context.getRegion());
+        this.caller = caller;
+        this.expectedTrace = expectedTrace;
     }
 
-    /*
-    if (trace1.getSegments().size() != trace2.getSegments().size()) {
-      throw new BaseException(ExceptionCode.TRACE_SPAN_LIST_NOT_MATCHED);
+    @Override
+    public void validate() throws Exception {
+        // get stored trace
+        Map<String, Object> storedTrace = this.getStoredTrace();
+
+        // create trace id list to retrieve trace from x-ray service
+        String traceId = (String) storedTrace.get("trace_id");
+        List<String> traceIdList = Collections.singletonList(traceId);
+
+        // get retrieved trace from x-ray service
+        Map<String, Object> retrievedTrace = this.getRetrievedTrace(traceIdList);
+
+        // validation of trace id
+        if (!storedTrace.get("trace_id").equals(retrievedTrace.get("trace_id"))) {
+            throw new BaseException(ExceptionCode.TRACE_ID_NOT_MATCHED);
+        }
+
+        // data model validation of other fields of segment document
+        for (Map.Entry<String, Object> entry : storedTrace.entrySet()) {
+            if (!entry.getValue().equals(retrievedTrace.get(entry.getKey()))) {
+                log.info("mis matched data model field list");
+                log.info("value of stored trace map: {}", entry.getValue());
+                log.info("value of retrieved map: {}",retrievedTrace.get(entry.getKey()));
+                log.info("==========================================");
+                throw new BaseException(ExceptionCode.DATA_MODEL_NOT_MATCHED);
+            }
+        }
+    }
+
+    // this method will hit get trace from x-ray service and get retrieved trace
+    private Map<String, Object> getRetrievedTrace(List<String> traceIdList) throws Exception {
+        Map<String, Object> flattenedJsonMapForRetrievedTrace = null;
+
+        // retrieve trace from x-ray service
+        TimeUnit.SECONDS.sleep(15);
```

Comment:
sleeping 15 seconds might not be enough in some case, instead, retry it until success could make this validation more stable.

Reply:
Yes initially I was trying retry only but with retry it doesn't block consecutive operation for example, it tried for the first time and `retrieveTraceList` is empty so it tries again but when it retries second time it actually executes program with whatever it got on first retry so I want something blocking in order to use retries. Basically, can we block retry handler until `retrieveTraceList` is populated and once we have the list then only it performs further execution.

Reply:
sorry i don't quite follow, can you retry on "listTraceByIds" until its size becomes none-zero?

Reply:
yes so the issue is retry helper retries when there is an exception but in the case of `listTraceByIds` we want to keep retrying until list becomes non empty. Here we have to wait for couple seconds because sample app sends trace to x-ray service and then we can retrieve it.

Reply:
i see, can you throw a baseexception when the list is empty so that it will retry